social_navigation_layers/src/human_layer.cpp

```diff
@@ -0,0 +1,223 @@
+//
+// This cost function is circular. From the center, to the lethal radius
+// the cost is "lethal." From there to the inscribed radius, the cost is
+// the "inscribed" cost. Beyond the inscribed radius, the cost gradually
+// decays exponentially.
+//
+// Args:
+//  (x, y): Coordinates of cost to be calculated
+//  (x0, y0): Coordinates of center of obstacle
+//  variance: Exponential variance for costs oustide of inscribed radius
+//            Increasing variance causes a slower decay of cost and a
+//            wider cirle of cost around a person.
+//  lethal_radius: Radius of lethal cost region (meters)
+//  inscribed_radius: Radius of inscribed cost region (meters)
+//
+// Returns:
+//  costmap cost, ranging from 0 to 254
+double HumanLayer::calculate_cost(double x, double y, double x0, double y0, double variance,
+                                  double lethal_radius, double inscribed_radius)
+{
+  double cost;
+  double dx = x - x0;
+  double dy = y - y0;
+  double radius = sqrt(dx * dx + dy * dy);
+  if (radius < lethal_radius)
+    cost = lethal_cost_;
+  else if (radius < inscribed_radius)
+    cost = inscribed_cost_;
+  else {
+    cost = inscribed_cost_ * exp(-(radius - inscribed_radius) / variance);
+    if (cost > inscribed_cost_)
+      cost = inscribed_cost_;
+  }
+
+  return cost;
+}
+
+// Calculates the maximum radius of costs for a human
+//
+// The cost function is exponential, which means it technically extends
+// to inifinite, but practically, any changes lower than the cutoff are
+// simply ignored. Calculate that maximum radius of pixels that need to
+// change surrounding an obstacle for a given cutoff and variance.
+//
+// Args:
+//  inscribed_radius: Radius of inscribed cost region (meters)
+//  cutoff: Cost threshold below which changes are not made to costmap
+//  variance: Exponential variance of cost function
+//
+// Returns:
+//  maximum radius of changes to costmap surrounding a human in meters
+double HumanLayer::get_radius(double inscribed_radius, double cutoff, double variance)
+{
+  return inscribed_radius - variance * log(cutoff / inscribed_cost_);
+}
+
+// Calculates the min/max bounding box of costmap changes for all people
+//
+// This is called by the costmap via the updateBounds method in the
+// parent SocialLayer class. Instead of updating all of the cells for
+// the entire costmap, only update a region of interest. This iterates
+// through all of the people detected and finds the minimum bounding box
+// to contain all of the changes for this update cycle.
+//
+// Output:
+//  (min_x, min_y, max_x, max_y) - Pointers to bounding box of changes
+void HumanLayer::updateBoundsFromPeople(double* min_x, double* min_y, double* max_x, double* max_y)
+{
+  double point = get_radius(inscribed_radius_, cutoff_, variance_);
+  std::list<people_msgs::Person>::iterator p_it;
+  for (p_it = transformed_people_.begin(); p_it != transformed_people_.end(); ++p_it)
+  {
+    people_msgs::Person person = *p_it;
```

Comment:
The compiler might figure out it can eliminate this copy, but might as well just make it explicit:
```
const people_msgs::Person& person = *p_it;
```

Reply:
Done.